### ISSUE TITLE
Support for Java modules

### ISSUE THREAD

**Issue by tdegueul** [alien-tools/roseau#8]
Java modules _should_ be the primary way for maintainers to specify the scope of their APIs and encapsulate internal packages since JDK9 circa 2017.

We've never implemented support for Java modules in Maracas nor Roseau but that should be top priority now. It took a while but adoption is increasing ([see some projects here](https://github.com/search?q=path%3A**%2Fmodule-info.java+exports&type=code&ref=advsearch)).

**Comment by tdegueul** [comment]
Some WIP notes after reading the JLS 21 (@jrfaller):

  - Any package that's not `exported` isn't part of the API, so free of compatibility guarantees
  - `exports X to Y` (_qualified_ export) limits the visibility of a package `X` to another "friend" module `Y`. So, there is _some_ module `Y` that might expect compatibility guarantees from `X`. Not sure how to handle that: is `X` API or not?
  - `opens`/`open module` enable reflective/run-time access, but not compile-time access; `opens` does not imply `exports`; so these should be irrelevant for us
  - `uses`/`requires` (and their variants) declare dependencies and are irrelevant
  - Services shouldn't be of interest for us: `provides ... with` etc.; ignore for now
  - Any compilation unit that's not part of a named module is part of the _unnamed module_. The unnamed module exports every package. No particular treatment needed.
  
[Chapter 13 of the JLS 21](https://docs.oracle.com/javase/specs/jls/se21/jls21.pdf#%5B%7B%22num%22%3A3651%2C%22gen%22%3A0%7D%2C%7B%22name%22%3A%22XYZ%22%7D%2C72%2C590%2Cnull%5D) expands a bit on _binary_ compatibility at the module level:

> If a module that was declared to export or open a package is changed to not export or open the package, or to export or open the package to a different set of friends, then an IllegalAccessError is thrown if a pre-existing binary is linked that needs but no longer has access to the public and protected classes and interfaces of the package. Such a change is not recommended for modules that have been widely distributed.
> 
> If a module was not declared to export or open a given package, then changing the module to export or open the package does not break compatibility with pre-existing binaries. However, changing the module to export the package may prevent the program from starting, since any module that reads the module may also read some other module that exports a package with the same name.
> 
> Adding a requires directive to a module declaration, or adding the transitive modifier to a requires directive, does not break compatibility with pre-existing binaries. However, it may prevent the program from starting, since the module may now read multiple modules that export packages with the same name.
> 
> Deleting a requires directive in a module declaration, or deleting the transitive modifier from a requires directive, may break compatibility with any pre-existing binary that relied on the directive or modifier for readability of a given module in the course of referencing classes and interfaces exported by that module. An IllegalAccessError may be thrown when such a reference from a pre-existing binary is linked.
> 
> Adding or deleting a uses or provides directive in a module declaration does not break compatibility with pre-existing binaries.

TL;DR: We might only be interested in `exports` directives, taking the unnamed module into account. From the initial API we infer, packages that are not explicitly exported should not be part of the API.
